angular/authentication/README.md

```diff
@@ -0,0 +1,130 @@
+# Authentication Guidelines
+
+## Refresh Token
+
+### Main Component
+
+By default, backend will set default token expiration time = 1 month. So, On frontend, The easiest way is just refreshing token every time when user refresh website. You can do by add code below in your `main.component.ts`
+
+```ts
+export class MainComponent implements OnInit {
+    ngOnInit() {
+        // call refresh token API
+        this.authService.refreshToken({refresh: localStorage.refresh}).subscribe(
+            (data: Token) => {
+                // save token and user information to local storage.
+                localStorage.setItem('currentUser', data);
+            },
+            (error) => {
+                // if error occurs, you may use old token and do nothing.
+                return;
+
+                // or redirect to login page.
+                // this.router.navigate(['/login']);
+            }
+        );
+    }
+}
+```
+
+### (Optional) Short Expiration Time
+
+If token time is very short (Ex. Banking website which token is expired in 5 minutes), You must check token expiration time before call request and refresh token if needed. See code below.
+
+```ts
+function refreshTokenIfExpired () {
+    // decode jwt token. This is very technical, you can skip this part of code.
+    let base64Url = localStorage.currentUser.split('.')[1];
+    let base64 = base64Url.replace(/-/g, '+').replace(/_/g, '/');
+    let jsonPayload = decodeURIComponent(atob(base64).split('').map(function(c) {
+        return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
+    }).join(''));
+
+    let decoded = JSON.parse(jsonPayload);
+    const now = Date.now().valueOf() / 1000
+
+    let refreshToken = false
+    // check if token will be expired in 5 minutes (300000 = 5 minutes)
+    let FIVE_MINUTES = 5 * 60 * 1000;
+    if (typeof decoded.exp !== 'undefined' && decoded.exp < now - FIVE_MINUTES) {
+        refreshToken = true
+    }
+    if (typeof decoded.nbf !== 'undefined' && decoded.nbf > now - FIVE_MINUTES) {
+        refreshToken = true
+    }
+
+    // call refresh token if it is nearly expired.
+    if (refreshToken) {
+        this.authService.refreshToken( ... )
+    }
+};
+```
+
+# Authentication Token Interceptor
+
+Authentication token interceptor which will add token to request.
+
+```ts
+import {Injectable} from '@angular/core';
+import {HttpEvent, HttpHandler, HttpInterceptor, HttpRequest} from '@angular/common/http';
+import {Observable} from 'rxjs';
+
+import {environment} from '../../../environments/environment';
+
+@Injectable()
+export class JwtInterceptor implements HttpInterceptor {
+  constructor(private authenticationService: AuthenticationService) {
+  }
+
+  intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
```

Comment:
Handle for some api that is not required token krub, I would recommend to follow this one https://stackoverflow.com/a/55522787/2306808